iroh/tests/gc.rs

```diff
@@ -0,0 +1,440 @@
+#![cfg(all(feature = "mem-db", feature = "iroh-collection"))]
+use std::{
+    io::{self, Cursor},
+    path::PathBuf,
+    time::Duration,
+};
+
+use anyhow::Result;
+use bao_tree::{
+    blake3,
+    io::{
+        fsm::{BaoContentItem, Outboard, ResponseDecoderReadingNext},
+        Leaf, Parent,
+    },
+    ChunkRanges,
+};
+use bytes::Bytes;
+use iroh::{baomap, node::Node};
+use iroh_io::AsyncSliceWriter;
+use rand::RngCore;
+use testdir::testdir;
+
+use iroh_bytes::{
+    baomap::{EntryStatus, Map, PartialMap, PartialMapEntry, Store, TempTag},
+    collection::LinkSeq,
+    util::{runtime, BlobFormat, HashAndFormat, Tag},
+    IROH_BLOCK_SIZE,
+};
+
+/// Pick up the tokio runtime from the thread local and add a
+/// thread per core runtime.
+fn test_runtime() -> runtime::Handle {
+    runtime::Handle::from_current(1).unwrap()
+}
+
+fn create_test_data(n: usize) -> Bytes {
+    let mut rng = rand::thread_rng();
+    let mut data = vec![0; n];
+    rng.fill_bytes(&mut data);
+    data.into()
+}
+
+/// Wrap a bao store in a node that has gc enabled.
+async fn wrap_in_node<S>(
+    bao_store: S,
+    rt: iroh_bytes::util::runtime::Handle,
+) -> Node<S, iroh_sync::store::memory::Store>
+where
+    S: iroh_bytes::baomap::Store,
+{
+    let doc_store = iroh_sync::store::memory::Store::default();
+    Node::builder(bao_store, doc_store)
+        .runtime(&rt)
+        .gc_policy(iroh::node::GcPolicy::Interval(Duration::from_millis(50)))
+        .spawn()
+        .await
+        .unwrap()
+}
+
+async fn gc_test_node() -> (
+    Node<baomap::mem::Store, iroh_sync::store::memory::Store>,
+    baomap::mem::Store,
+) {
+    let rt = test_runtime();
+    let bao_store = baomap::mem::Store::new(rt.clone());
+    let node = wrap_in_node(bao_store.clone(), rt).await;
+    (node, bao_store)
+}
+
+async fn step() {
+    tokio::time::sleep(Duration::from_millis(100)).await;
```

Comment:
why not use tokios timer https://docs.rs/tokio/latest/tokio/time/fn.pause.html#auto-advance testing functionality, this actually worked quite well for me when testing other timer based behaviour?

Reply:
I hate it.

It depends on the notion that every single place in your code base and your dependencies that deals with time is using the tokio stuff, otherwise it will break in weird ways. Have a single place in your own code or your library code where you compute differences between `SystemTime`, and it is going to fail. But not outright fail. Fail in a subtle way.

It's another of these attempts of the tokio people to make tokio the default stdlib of rust. Is Lennart Poettering of systemd a secret tokio contributor?

I had thought about having an event stream from the db that you could use to detect when a GC has been performed instead of a fixed delay. I did not do it bc I thought that would lock in the fact that we are using a certain GC method (mark and sweep). But I would prefer to do this over using that tokio stuff.

Reply:
> Is Lennart Poettering of systemd a secret tokio contributor?

🤣